contracts/account/paymaster/PaymasterERC20.sol

```diff
@@ -0,0 +1,115 @@
+            userOp.maxFeePerGas()).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        return (
+            abi.encodePacked(userOpHash, token, prefundAmount, tokenPrice, userOp.sender, guarantor),
+            token
+                .trySafeTransferFrom(guarantor == address(0) ? userOp.sender : guarantor, address(this), prefundAmount)
+                .packValidationData(validAfter, validUntil)
+        );
+    }
+
+    /// @inheritdoc PaymasterCore
+    function _postOp(
+        PostOpMode /* mode */,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) internal virtual override {
+        bytes32 userOpHash = bytes32(context[0x00:0x20]);
+        IERC20 token = IERC20(address(bytes20(context[0x20:0x34])));
+        uint256 prefundAmount = uint256(bytes32(context[0x34:0x54]));
+        uint256 tokenPrice = uint256(bytes32(context[0x54:0x74]));
+        address user = address(bytes20(context[0x74:0x88]));
+        address guarantor = address(bytes20(context[0x88:0x9C]));
+
+        uint256 actualAmount = (actualGasCost +
+            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
+            actualUserOpFeePerGas).mulDiv(tokenPrice, _tokenPriceDenominator());
+
+        if (guarantor == address(0)) {
+            token.safeTransfer(user, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, address(0), actualAmount, tokenPrice, false);
+        } else if (token.trySafeTransferFrom(user, address(this), actualAmount)) {
+            token.safeTransfer(guarantor, prefundAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, false);
+        } else {
+            token.safeTransfer(guarantor, prefundAmount - actualAmount);
+            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, true);
+        }
+    }
+
+    /**
+     * @dev Internal function that returns the repayment details for a given user operation
+     *
+     * This may be implemented in any number of ways, including
+     * * Hardcoding values (only one token supported)
+     * * Getting the price from an onchain oracle
+     * * Getting the (signed) values through the userOp's paymasterData
+     *
+     * The paymaster can also decide to not support guarantors, and always return address(0) for that part.
+     */
+    function _fetchDetails(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash
+    )
+        internal
+        view
+        virtual
+        returns (IERC20 token, uint48 validAfter, uint48 validUntil, uint256 tokenPrice, address guarantor);
+
+    /// @dev Denominator used for interpreting the `tokenPrice` returned by {_fetchDetails} as "fixed point".
+    function _tokenPriceDenominator() internal view virtual returns (uint256) {
+        return 1e18;
```

Comment:
Isn't denominator too high? For example, if the `maxFeePerGas` is 1e9, then any price less than 1e9 will be zeroed out.

Am I missing something?

Reply:
It used to be 1e6, then @arr00 said it was too low and 1e18 would be better. This is virtual, so devs can override it.

Note that this is used in `Math.mulDiv` to "scale" the token price (in native). So you could put anything here, it would just re-scale the token price. I'm not sure how `maxFeePerGas` has an impact here.